# PRAKTIKUM 9

## Python Learning

In [1]:
print("Hello, World!")

Hello, World!


### Analisis

Kode print("Hello, World!") digunakan untuk menampilkan teks Hello, World! di layar.
Program ini merupakan contoh paling sederhana dalam Python dan biasanya digunakan sebagai langkah awal dalam mempelajari pemrograman.
Pada dasarnya, fungsi print() berfungsi menampilkan pesan atau output kepada pengguna.

## Python Indentation

In [2]:
if 5 > 2:
  print("Five is greater than two!")

Five is greater than two!


### Analisis

Kode if 5 > 2: print("Five is greater than two!") adalah pernyataan kondisional dalam Python. Program memeriksa apakah 5 lebih besar dari 2; karena pernyataan tersebut bernilai benar, maka teks "Five is greater than two!" akan ditampilkan di layar.


In [3]:
if 5 > 2:
 print("Five is greater than two!") 
if 5 > 2:
        print("Five is greater than two!") 

Five is greater than two!
Five is greater than two!


### Analisis

Kode tersebut berisi dua pernyataan if yang memeriksa kondisi yang sama, yaitu apakah 5 lebih besar dari 2. Karena kondisi tersebut benar, kedua pernyataan akan dijalankan dan menghasilkan output yang sama dua kali.

## Python Variables and Comments

In [3]:
x = 5
y = "Hello, World!"

In [4]:
#This is a comment.
print("Hello, World!")

Hello, World!


### Analisis

Kode ini menetapkan dua variabel, yaitu x dengan nilai 5 dan y dengan nilai "Hello, World!".
Tanda #This is a comment merupakan komentar yang tidak dijalankan oleh program.
Pada baris terakhir, perintah print("Hello, World!") berfungsi untuk menampilkan teks Hello, World! pada layar.

## Statements

In [5]:
print("Python is fun!")

Python is fun!


### Analisis

Perintah print("Python is fun!") digunakan untuk menampilkan teks Python is fun! di layar.
Pernyataan ini menggambarkan cara dasar Python dalam menjalankan perintah untuk menampilkan hasil kepada pengguna.

## Many Statements

In [4]:
print("Hello World!")
print("Have a good day.")
print("Learning Python is fun!")

Hello World!
Have a good day.
Learning Python is fun!


### Analisis

Kode ini terdiri dari tiga perintah print() yang dieksekusi secara berurutan untuk menampilkan tiga teks berbeda di layar.
Hal ini memperlihatkan bahwa Python menjalankan setiap perintah berdasarkan urutan penulisan dalam kode.

## Semicolons (Optional, Rarely Used)

In [5]:
print("Hello"); print("How are you?"); print("Bye bye!")

Hello
How are you?
Bye bye!


### Analisis

Kode ini memakai tanda titik koma (;) untuk memisahkan beberapa perintah print() dalam satu baris sehingga semuanya dijalankan secara berurutan. Meskipun berfungsi, cara ini jarang dipakai karena dapat mengurangi keterbacaan kode. Sebaiknya, setiap perintah ditulis pada baris terpisah agar lebih jelas.

## Print Text

In [8]:
print("Hello World!")

Hello World!


### Analisis

Kode print("Hello World!") berfungsi untuk menampilkan teks ke layar. Program ini memperlihatkan penggunaan dasar fungsi print() dalam Python untuk menghasilkan output berupa tulisan Hello World!.

In [9]:
print("Hello World!")
print("I am learning Python.")
print("It is awesome!")

Hello World!
I am learning Python.
It is awesome!


### Analisis

Kode ini terdiri dari tiga perintah print() yang masing-masing menampilkan teks berbeda secara berurutan. Akibatnya, program akan menghasilkan tiga baris output di layar, yaitu Hello World!, I am learning Python., dan It is awesome!.

## Double Quotes

In [10]:
print("This will work!")
print('This will also work!')

This will work!
This will also work!


### Analisis

Kode ini memperlihatkan bahwa Python dapat memakai tanda kutip ganda (" ") maupun tanda kutip tunggal (' ') untuk menuliskan teks. Kedua jenis kutip tersebut memiliki fungsi yang sama, sehingga perintah print("This will work!") dan print('This will also work!') akan menampilkan hasil identik di layar.

## Print Without a New Line

In [11]:
print("Hello World!", end=" ")
print("I will print on the same line.")

Hello World! I will print on the same line.


### Analisis

Kode ini memanfaatkan parameter end=" " pada fungsi print() agar output selanjutnya ditampilkan di baris yang sama. Dengan cara ini, program menghasilkan teks Hello World! I will print on the same line. dalam satu baris, bukan pada dua baris terpisah.

## Print Numbers

In [12]:
print(3)
print(358)
print(50000)

3
358
50000


### Analisis

Kode ini terdiri dari tiga perintah print() yang masing-masing menampilkan angka berbeda. Ketika dijalankan, program akan menampilkan angka 3, 358, dan 50000 secara berurutan pada tiga baris yang terpisah.

In [6]:
print(3 + 3)
print(2 * 5)

6
10


### Analisis

Kode ini menjalankan operasi aritmetika melalui fungsi print(). Pada baris pertama dilakukan penjumlahan 3 + 3 yang menghasilkan nilai 6, sedangkan pada baris kedua dilakukan perkalian 2 * 5 yang menghasilkan nilai 10.

## Mix Text and Numbers

In [14]:
print("I am", 35, "years old.")

I am 35 years old.


### Analisis

Kode ini memanfaatkan fungsi print() untuk menampilkan kombinasi teks dan angka secara bersamaan. Output yang dihasilkan di layar adalah I am 35 years old. dengan spasi otomatis di antara setiap elemen.

## Creating a Comment

In [15]:
#This is a comment
print("Hello, World!")

Hello, World!


### Analisis

Kode ini terdiri dari komentar dan perintah print(). Baris yang dimulai dengan tanda pagar (#) merupakan komentar yang tidak dijalankan oleh Python, sedangkan baris print("Hello, World!") berfungsi untuk menampilkan teks Hello, World! ke layar.

In [16]:
print("Hello, World!") #This is a comment

Hello, World!


### Analisis

Kode ini menampilkan teks Hello, World! di layar. Sementara itu, bagian setelah tanda pagar (#This is a comment) merupakan komentar yang diabaikan oleh Python dan berfungsi sebagai catatan atau penjelasan dalam kode.

In [17]:
#print("Hello, World!")
print("Cheers, Mate!")

Cheers, Mate!


### Analisis

Kode ini berisi satu baris yang diberi tanda pagar (#), sehingga dianggap sebagai komentar dan tidak dijalankan. Hanya baris print("Cheers, Mate!") yang dieksekusi, menghasilkan tampilan teks Cheers, Mate! di layar.

## Multiline Comments

In [18]:
#This is a comment
#written in
#more than just one line
print("Hello, World!")

Hello, World!


### Analisis

Kode ini terdiri dari beberapa baris komentar yang dimulai dengan tanda pagar (#) dan tidak dieksekusi oleh Python. Hanya baris print("Hello, World!") yang dijalankan untuk menampilkan teks Hello, World! di layar.

In [19]:
"""
This is a comment
written in
more than just one line
"""
print("Hello, World!")

Hello, World!


### Analisis

Kode ini memakai tanda kutip tiga (""" ... """) untuk menuliskan komentar atau teks dalam beberapa baris yang diabaikan oleh Python. Hanya perintah print("Hello, World!") yang dijalankan, sehingga menghasilkan output Hello, World! di layar.

## Creating Variables

In [20]:
x = 5
y = "John"
print(x)
print(y)

5
John


### Analisis

Kode ini mendefinisikan dua variabel, yaitu x bernilai 5 dan y bernilai "John". Selanjutnya, fungsi print() digunakan untuk menampilkan nilai dari kedua variabel tersebut.

In [7]:
x = 4       # x is of type int
x = "Sally" # x is now of type str
print(x)


Sally


### Analisis

Kode ini memperlihatkan bahwa di Python, nilai dan tipe data suatu variabel dapat berubah secara dinamis. Pada awalnya, x bernilai 4, kemudian diubah menjadi "Sally", sehingga hasil yang ditampilkan adalah Sally.

## Casting

In [22]:
x = str(3)    # x will be '3'
y = int(3)    # y will be 3
z = float(3)  # z will be 3.0

### Analisis

Kode ini memperlihatkan bagaimana melakukan konversi tipe data di Python. Variabel x dikonversi menjadi string ('3'), y menjadi bilangan bulat (3), dan z menjadi bilangan desimal (3.0).

## Get the Type

In [23]:
x = 5
y = "John"
print(type(x))
print(type(y))

<class 'int'>
<class 'str'>


### Analisis

Kode ini menggunakan fungsi type() untuk menampilkan tipe data dari dua variabel. Hasilnya menunjukkan bahwa x memiliki tipe data int (bilangan bulat), sedangkan y bertipe str (teks atau string).

## Single or Double Quotes

In [8]:
x = "John"
# is the same as
x = 'John'

### Analisis

Kode ini memperlihatkan bahwa Python dapat memakai tanda kutip tunggal (' ') maupun tanda kutip ganda (" ") untuk membuat sebuah string. Kedua cara tersebut memiliki fungsi yang sama dan menghasilkan nilai John.

## Case-Sensitive

In [9]:
a = 4
A = "Sally"
#A will not overwrite a

### Analisis

Kode ini memperlihatkan bahwa Python bersifat case-sensitive, artinya huruf besar dan kecil pada nama variabel dibedakan. Variabel a memiliki nilai 4, sedangkan A bernilai "Sally", dan keduanya dianggap sebagai variabel yang berbeda.

## Variable Names

In [10]:
myvar = "John"
my_var = "John"
_my_var = "John"
myVar = "John"
MYVAR = "John"
myvar2 = "John"

### Analisis

Kode ini menunjukkan beberapa contoh penulisan nama variabel yang sah dalam Python. Semua contoh tersebut valid karena terdiri dari huruf, angka, dan garis bawah, serta tidak diawali dengan angka.

## Camel Case

In [28]:
myVariableName = "John"

### Analisis

Kode ini menerapkan gaya penulisan camelCase, yaitu huruf pertama pada kata pertama ditulis kecil, sedangkan setiap kata berikutnya diawali huruf besar. Gaya ini umum digunakan untuk penamaan variabel, misalnya myVariableName.

## Pascal Case

In [11]:
MyVariableName = "John"

### Analisis

Kode ini menggunakan gaya penulisan PascalCase, yaitu setiap kata diawali dengan huruf besar. Gaya ini biasanya digunakan untuk penamaan kelas dalam Python, seperti pada contoh MyVariableName.

## Snake Case

In [30]:
my_variable_name = "John"

### Analisis

Kode ini menerapkan gaya penulisan snake_case, di mana setiap kata dipisahkan dengan tanda garis bawah (_). Gaya ini sering digunakan dalam Python untuk menamai variabel maupun fungsi.

## Many Values to Multiple Variables

In [31]:
x, y, z = "Orange", "Banana", "Cherry"
print(x)
print(y)
print(z)

Orange
Banana
Cherry


### Analisis

Kode ini memanfaatkan teknik multiple assignment untuk memberikan nilai pada tiga variabel secara bersamaan. Variabel x bernilai "Orange", y bernilai "Banana", dan z bernilai "Cherry". Ketika dijalankan, program menampilkan ketiga nilai tersebut pada baris yang terpisah.

## One Value to Multiple Variables

In [33]:
x = y = z = "Orange"
print(x)
print(y)
print(z)

Orange
Orange
Orange


### Analisis

Kode ini menetapkan nilai yang sama, yaitu "Orange", pada tiga variabel (x, y, dan z) sekaligus. Ketika program dijalankan, hasil yang ditampilkan adalah teks Orange sebanyak tiga kali pada baris yang terpisah.

## Unpack a Collection

In [34]:
fruits = ["apple", "banana", "cherry"]
x, y, z = fruits
print(x)
print(y)
print(z)

apple
banana
cherry


### Analisis

Kode ini melakukan unpacking terhadap sebuah daftar (list) ke dalam tiga variabel. Nilai "apple" disimpan pada x, "banana" pada y, dan "cherry" pada z. Ketika dijalankan, program menampilkan ketiga nilai tersebut pada baris yang berbeda.

## Output Variables

In [35]:
x = "Python is awesome"
print(x)


Python is awesome


### Analisis

Kode ini menyimpan teks "Python is awesome" ke dalam variabel x, lalu menampilkannya di layar menggunakan fungsi print().

In [36]:
x = "Python"
y = "is"
z = "awesome"
print(x, y, z)

Python is awesome


### Analisis

Kode ini mendefinisikan tiga variabel yang masing-masing berisi teks berbeda, kemudian mencetaknya secara bersamaan dengan spasi otomatis di antara tiap kata. Output yang dihasilkan adalah Python is awesome.

In [37]:
x = "Python "
y = "is "
z = "awesome"
print(x + y + z)

Python is awesome


### Analisis

Kode ini menggabungkan tiga string menggunakan operator +. Output yang dihasilkan adalah Python is awesome dalam satu baris, tanpa spasi tambahan selain yang terdapat di dalam variabel.

In [12]:
x = 5
y = 10
print(x + y)

15


### Analisis

Kode ini digunakan untuk menjumlahkan dua variabel, yaitu x = 5 dan y = 10, kemudian menampilkan hasilnya di layar. Perintah print(x + y) melakukan operasi penjumlahan antara kedua nilai tersebut. Saat dijalankan, program menghasilkan output 15 sebagai hasil dari operasi 5 + 10.

In [13]:
x = 5
y = "John"
print(x, y)

5 John


### Analisis

Kode ini menetapkan nilai 5 pada variabel x dan "John" pada variabel y. Perintah print(x, y) digunakan untuk menampilkan kedua nilai tersebut secara bersamaan di layar, sehingga output yang dihasilkan adalah 5 John.

## Global Variables

In [41]:
x = "awesome"

def myfunc():
  print("Python is " + x)

myfunc()

Python is awesome


### Analisis

Kode tersebut menetapkan nilai "awesome" ke dalam variabel x dan membuat fungsi myfunc() yang mencetak teks hasil penggabungan dengan nilai variabel tersebut. Saat myfunc() dijalankan, program akan menampilkan output "Python is awesome" di layar.

In [42]:
x = "awesome"

def myfunc():
  x = "fantastic"
  print("Python is " + x)

myfunc()

print("Python is " + x)

Python is fantastic
Python is awesome


### Analisis

Kode tersebut mendefinisikan variabel global x = "awesome" dan variabel lokal x = "fantastic" di dalam fungsi myfunc(). Saat fungsi dijalankan, program menampilkan "Python is fantastic" karena menggunakan variabel lokal. Setelah fungsi selesai, perintah di luar fungsi mencetak "Python is awesome", yang berasal dari variabel global.

## The global Keyword

In [43]:
def myfunc():
  global x
  x = "fantastic"

myfunc()

print("Python is " + x)

Python is fantastic


### Analisis

Kode tersebut menggunakan kata kunci global agar variabel x dapat diakses dan diubah di luar fungsi. Setelah myfunc() dijalankan, nilai variabel x menjadi "fantastic", sehingga output yang ditampilkan adalah "Python is fantastic".

In [44]:
x = "awesome"

def myfunc():
  global x
  x = "fantastic"

myfunc()

print("Python is " + x)

Python is fantastic


### Analisis

Kode tersebut mula-mula mendefinisikan variabel x = "awesome", lalu di dalam fungsi myfunc() digunakan kata kunci global untuk memungkinkan perubahan nilai x di luar fungsi. Setelah fungsi dijalankan, nilai x berubah menjadi "fantastic", sehingga output program yang dihasilkan adalah "Python is fantastic".

## Getting the Data Type

In [14]:
x = 5
print(type(x))

<class 'int'>


### Analisis 

Kode tersebut menetapkan nilai 5 pada variabel x, kemudian menggunakan fungsi type() untuk menampilkan jenis tipe datanya. Saat program dijalankan, hasil yang ditampilkan adalah <class 'int'>, yang menunjukkan bahwa variabel x bertipe integer atau bilangan bulat.

## Python Numbers

In [46]:
x = 1    # int
y = 2.8  # float
z = 1j   # complex

### Analisis

Kode tersebut mendefinisikan tiga variabel dengan tipe data numerik yang berbeda. Variabel x = 1 bertipe int (bilangan bulat), y = 2.8 bertipe float (bilangan desimal), dan z = 1j bertipe complex (bilangan kompleks dengan bagian imajiner). Ketiga variabel tersebut merepresentasikan tiga jenis utama tipe data numerik dalam Python.

In [16]:
print(type(x))
print(type(y))
print(type(z))

<class 'int'>
<class 'str'>


NameError: name 'z' is not defined

### Analisis

Kode tersebut memanfaatkan fungsi type() untuk menampilkan tipe data dari setiap variabel. Fungsi ini digunakan untuk memastikan bahwa x, y, dan z memiliki tipe data yang berbeda, yaitu int, float, dan complex.

## Int

In [48]:
x = 1
y = 35656222554887711
z = -3255522

print(type(x))
print(type(y))
print(type(z))

<class 'int'>
<class 'int'>
<class 'int'>


### Analisis

Kode tersebut mendefinisikan tiga variabel bilangan bulat dengan nilai yang berbeda, yaitu x = 1, y = 35656222554887711, dan z = -3255522. Selanjutnya, fungsi type() digunakan untuk menampilkan tipe data dari masing-masing variabel, yang semuanya bertipe int (bilangan bulat) dalam Python.

## Float

In [49]:
x = 1.10
y = 1.0
z = -35.59

print(type(x))
print(type(y))
print(type(z))

<class 'float'>
<class 'float'>
<class 'float'>


### Analisis 

Kode tersebut mendefinisikan tiga variabel dengan nilai desimal, yaitu x = 1.10, y = 1.0, dan z = -35.59. Ketiga variabel tersebut termasuk dalam tipe data float, karena masing-masing memiliki tanda desimal yang menunjukkan bilangan pecahan.

In [50]:
x = 35e3
y = 12E4
z = -87.7e100

print(type(x))
print(type(y))
print(type(z))

<class 'float'>
<class 'float'>
<class 'float'>


### Analisis

Kode tersebut menggunakan notasi eksponensial (scientific notation) untuk merepresentasikan nilai desimal yang sangat besar dalam variabel. Nilai x = 35e3 setara dengan 35 × 10³ = 35000, y = 12E4 setara dengan 12 × 10⁴ = 120000, dan z = -87.7e100 setara dengan -87.7 × 10¹⁰⁰. Ketiganya bertipe data float dalam Python.

## Complex

In [51]:
x = 3+5j
y = 5j
z = -5j

print(type(x))
print(type(y))
print(type(z))

<class 'complex'>
<class 'complex'>
<class 'complex'>


### Analisis

Kode tersebut mendefinisikan tiga variabel dengan tipe bilangan kompleks, yaitu x = 3+5j, y = 5j, dan z = -5j. Huruf j digunakan menunjukkan bagian imajiner dalam Python.

## Type Conversion

In [52]:
x = 1    # int
y = 2.8  # float
z = 1j   # complex

#convert from int to float:
a = float(x)

#convert from float to int:
b = int(y)

#convert from int to complex:
c = complex(x)

print(a)
print(b)
print(c)

print(type(a))
print(type(b))
print(type(c))

1.0
2
(1+0j)
<class 'float'>
<class 'int'>
<class 'complex'>


### Analisis 

Kode tersebut menunjukkan proses konversi tipe data numerik. Variabel x, y, dan z masing-masing berisi nilai int, float, dan complex. Nilai x dikonversi menjadi float disimpan pada variabel a, menghasilkan nilai 1.0. Selanjutnya, nilai y dikonversi menjadi int disimpan pada variabel b, menghasilkan nilai 2 karena bagian desimalnya dihilangkan. Terakhir, nilai x dikonversi menjadi complex dan disimpan pada variabel c, menghasilkan (1+0j). Saat program dijalankan, output yang tampilkan menunjukkan hasil konversi beserta tipe datanya, yaitu float, int, dan complex.

## Random Number

In [53]:
import random

print(random.randrange(1, 10))


9


### Analisis

Kode tersebut menggunakan modul random untuk menghasilkan angka acak. Perintah random.randrange(1, 10) akan menampilkan satu angka acak antara 1 hingga 9, karena batas atas (10) tidak termasuk dalam hasil. Setiap kali program dijalankan, angka yang muncul bisa berbeda-beda.

## Specify a Variable Type

In [54]:
x = int(1)   # x will be 1
y = int(2.8) # y will be 2
z = int("3") # z will be 3

### Analisis

Kode tersebut memperlihatkan proses konversi tipe data ke integer menggunakan fungsi int(). Variabel x berasal dari bilangan bulat 1, y dari bilangan desimal 2.8 yang dibulatkan ke bawah menjadi 2, dan z dari string "3" yang dikonversi menjadi angka 3. Setelah dikonversi, ketiga variabel tersebut memiliki tipe data int dengan nilai akhir 1, 2, dan 3.

In [55]:
x = float(1)     # x will be 1.0
y = float(2.8)   # y will be 2.8
z = float("3")   # z will be 3.0
w = float("4.2") # w will be 4.2

### Analisis

Kode tersebut melakukan konversi tipe data ke float menggunakan fungsi float(). Variabel x dikonversi dari bilangan bulat menjadi 1.0, y sudah berupa 2.8 sehingga tetap sama, z dikonversi dari string "3" menjadi 3.0, dan w dari string "4.2" menjadi 4.2. Hasil akhirnya, semua variabel memiliki nilai desimal dan bertipe float.

In [17]:
x = str("s1") # x will be 's1'
y = str(2)    # y will be '2'
z = str(3.0)  # z will be '3.0'

### Analisis 

Kode tersebut melakukan konversi nilai ke tipe data string menggunakan fungsi str(). Variabel x diubah menjadi '1', y menjadi '2', dan z menjadi '3.0'. Setelah dikonversi, semua nilai disimpan dalam bentuk teks (string) meskipun sebelumnya memiliki tipe data yang berbeda.

## Strings

In [58]:
print("Hello")
print('Hello')

Hello
Hello


### Analisis

Kode tersebut menampilkan teks "Hello" ke layar. Dalam Python, penggunaan tanda kutip ganda (" ") maupun tanda kutip tunggal (' ') memiliki fungsi yang sama, yaitu untuk mendefinisikan sebuah string.

## Quotes Inside Quotes

In [59]:
print("It's alright")
print("He is called 'Johnny'")
print('He is called "Johnny"')

It's alright
He is called 'Johnny'
He is called "Johnny"


### Analisis

Kode tersebut menampilkan teks yang mengandung tanda kutip di dalam string. Python memungkinkan penggunaan tanda kutip tunggal dan ganda secara fleksibel untuk menghindari konflik. Pada baris pertama digunakan tanda kutip ganda agar tanda apostrof pada kata It's tidak dianggap sebagai penutup string, sehingga teks dapat ditampilkan dengan benar.

## Assign String to a Variable

In [60]:
a = "Hello"
print(a)

Hello


### Analisis

Kode tersebut menyimpan string "Hello" ke dalam variabel a, kemudian menggunakan perintah print(a) untuk menampilkan isi variabel tersebut ke layar. Hasil output yang muncul adalah Hello.

## Multiline Strings

In [61]:
a = """Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua."""
print(a)

Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua.


### Analisis

Kode tersebut menyimpan teks multiline ke dalam variabel a dengan menggunakan tiga tanda kutip ganda (""" """"). Cara ini memungkinkan string mencakup beberapa baris tanpa memerlukan karakter khusus untuk baris baru. Saat perintah print(a) dijalankan, seluruh teks akan ditampilkan persis seperti penulisannya, termasuk jeda barisnya.

In [62]:
a = '''Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua.'''
print(a)

Lorem ipsum dolor sit amet,
consectetur adipiscing elit,
sed do eiusmod tempor incididunt
ut labore et dolore magna aliqua.


### Analisis

Kode tersebut menyimpan teks multiline ke dalam variabel a menggunakan tiga tanda kutip tunggal (''' '''). Cara ini sama seperti penggunaan tiga tanda kutip ganda, yang memungkinkan string mencakup beberapa baris secara langsung. Saat perintah print(a) dijalankan, seluruh teks akan ditampilkan persis seperti ditulis, termasuk pindah barisnya.

## Strings are Arrays

In [64]:
a = "Hello, World!"
print(a[1])

e


### Analisis

Kode tersebut menyimpan string "Hello, World!" ke dalam variabel a, kemudian menggunakan a[1] untuk mengakses karakter pada indeks ke-1. Karena dalam Python indeks dimulai dari 0, maka karakter pada posisi 1 adalah huruf 'e'.

## Looping Through a String

In [65]:
for x in "banana":
  print(x)

b
a
n
a
n
a


### Analisis

Kode tersebut menggunakan perulangan for untuk menelusuri setiap karakter dalam string "banana". Pada setiap iterasi, variabel x akan menyimpan satu karakter dari string tersebut, lalu karakter tersebut dicetak ke layar satu per satu.

## String Length

In [66]:
a = "Hello, World!"
print(len(a))

13


### Analisis

Kode tersebut menyimpan string "Hello, World!" ke dalam variabel a, kemudian menggunakan fungsi len() untuk menghitung jumlah total karakter di dalam string tersebut, termasuk spasi dan tanda baca.

## Check String

In [67]:
txt = "The best things in life are free!"
print("free" in txt)

True


### Analisis

Kode tersebut menyimpan string "The best things in life are free!" ke dalam variabel txt, kemudian menggunakan operator in untuk memeriksa apakah kata "free" terdapat di dalam string tersebut. Jika benar, hasilnya akan bernilai True.

In [68]:
txt = "The best things in life are free!"
if "free" in txt:
  print("Yes, 'free' is present.")

Yes, 'free' is present.


### Analisis

Kode tersebut menggunakan pernyataan if untuk memeriksa apakah kata "free" terdapat di dalam string txt. Jika kondisi tersebut bernilai True, maka perintah print() akan dijalankan untuk menampilkan pesan yang telah ditentukan ke layar.

## Check if NOT

In [69]:
txt = "The best things in life are free!"
print("expensive" not in txt)

True


### Analisis

Kode tersebut menggunakan operator not in untuk memeriksa apakah kata "expensive" tidak terdapat dalam string txt. Karena kata tersebut memang tidak ada di dalam teks, maka hasilnya bernilai True, dan perintah print() akan menampilkan True ke layar.

In [70]:
txt = "The best things in life are free!"
if "expensive" not in txt:
  print("No, 'expensive' is NOT present.")

No, 'expensive' is NOT present.


### Analisis

Kode tersebut menggunakan operator not in untuk memeriksa apakah kata "expensive" tidak terdapat dalam string txt. Karena kata tersebut memang tidak ditemukan, maka kondisi bernilai True dan perintah print() dijalankan. Hasil output yang ditampilkan adalah No, 'expensive' is NOT present., menunjukkan bahwa Python dapat mengeksekusi perintah tertentu berdasarkan keberadaan atau ketiadaan teks dalam sebuah string.

## Slicing

In [71]:
b = "Hello, World!"
print(b[2:5])

llo


### Analisis

Kode tersebut menggunakan slicing untuk mengambil sebagian isi dari string b = "Hello, World!". Ekspresi b[2:5] akan mengambil karakter mulai dari indeks 2 hingga 4 (karena indeks 5 tidak termasuk), sehingga hasilnya adalah substring "llo".

## Slice From the Start

In [72]:
b = "Hello, World!"
print(b[:5])


Hello


### Analisis

Kode tersebut menggunakan slicing untuk mengambil sebagian isi dari string b = "Hello, World!". Ekspresi b[:5] mengambil karakter dari awal string (indeks 0) hingga indeks 4 (karena indeks 5 tidak termasuk), sehingga hasilnya adalah substring "Hello".

## Slice To the End

In [74]:
b = "Hello, World!"
print(b[2:])

llo, World!


### Analisis

Kode tersebut menggunakan slicing untuk mengambil sebagian isi dari string b = "Hello, World!". Ekspresi b[2:] akan mengambil semua karakter mulai dari indeks 2 hingga akhir string, sehingga hasilnya adalah substring "llo, World!".

## Negative Indexing

In [75]:
b = "Hello, World!"
print(b[-5:-2])

orl


### Analisis

Kode tersebut menggunakan slicing dengan indeks negatif pada string b = "Hello, World!". Ekspresi b[-5:-2] mengambil karakter mulai dari indeks -5 hingga -3 (karena indeks -2 tidak termasuk), sehingga hasilnya adalah substring "orl".

## Upper Case

In [76]:
a = "Hello, World!"
print(a.upper())

HELLO, WORLD!


### Analisis

Kode tersebut menggunakan metode upper() untuk mengubah seluruh huruf dalam string a = "Hello, World!" menjadi huruf kapital. Hasilnya, string tersebut akan berubah menjadi "HELLO, WORLD!".

## Lower Case

In [77]:
a = "Hello, World!"
print(a.lower())


hello, world!


### Analisis

Kode tersebut menggunakan metode lower() untuk mengubah seluruh huruf dalam string a = "Hello, World!" menjadi huruf kecil. Hasilnya, string tersebut akan menjadi "hello, world!".

## Remove Whitespace

In [79]:
a = " Hello, World! "
print(a.strip()) # returns "Hello, World!"

Hello, World!


### Analisis

Kode tersebut menggunakan metode strip() untuk menghapus spasi kosong di awal dan akhir string a = " Hello, World! ". Saat perintah print(a.strip()) dijalankan, hasil yang ditampilkan adalah "Hello, World!", yaitu string tanpa spasi di kedua sisinya.

## Replace String

In [80]:
a = "Hello, World!"
print(a.replace("H", "J"))

Jello, World!


### Analisis

Kode tersebut menggunakan metode replace() untuk mengganti karakter "H" dengan "J" pada string a = "Hello, World!". Saat perintah print(a.replace("H", "J")) dijalankan, hasil yang ditampilkan adalah "Jello, World!".

## Split String

In [81]:
a = "Hello, World!"
print(a.split(",")) # returns ['Hello', ' World!']

['Hello', ' World!']


### Analisis

Kode tersebut menggunakan metode split() untuk memisahkan string a = "Hello, World!" menjadi sebuah daftar (list) dengan pemisah berupa koma (","). Saat perintah print(a.split(",")) dijalankan, hasil yang ditampilkan adalah ['Hello', ' World!'].

## String Concatenation

In [82]:
a = "Hello"
b = "World"
c = a + b
print(c)

HelloWorld


### Analisis

Kode tersebut menggunakan operator + untuk menggabungkan (concatenate) dua string. Variabel a = "Hello" dan b = "World" digabung menjadi c, sehingga saat perintah print(c) dijalankan, hasil yang ditampilkan adalah "HelloWorld".

In [84]:
a = "Hello"
b = "World"
c = a + " " + b
print(c)

Hello World


### Analisis

Kode tersebut menggunakan operator + untuk menggabungkan dua string. Variabel a = "Hello" dan b = "World" digabung dengan menambahkan spasi di antara keduanya, misalnya c = a + " " + b. Saat perintah print(c) dijalankan, hasil yang ditampilkan adalah "Hello World".

## F-Strings

In [85]:
age = 36
txt = f"My name is John, I am {age}"
print(txt)

My name is John, I am 36


### Analisis

Kode tersebut menggunakan f-string di Python untuk menyisipkan nilai variabel age ke dalam sebuah string. Variabel age = 36 secara otomatis dimasukkan ke dalam teks pada posisi {age}, sehingga ketika dijalankan, hasil outputnya adalah "My name is John, and I am 36".

## Placeholders and Modifiers

In [86]:
price = 59
txt = f"The price is {price} dollars"
print(txt)

The price is 59 dollars


### Analisis

Kode tersebut menggunakan f-string di Python untuk menyisipkan nilai variabel age ke dalam sebuah string. Variabel age = 36 secara otomatis dimasukkan ke dalam teks pada posisi {age}, sehingga ketika dijalankan, hasil outputnya adalah "My name is John, and I am 36".

In [87]:
price = 59
txt = f"The price is {price:.2f} dollars"
print(txt)

The price is 59.00 dollars


### Analisis

Kode tersebut menggunakan f-string dengan format spesifikasi :.2f untuk menampilkan nilai variabel price = 59 dalam bentuk desimal dengan dua angka di belakang koma. Saat dijalankan, hasil outputnya adalah "The price is 59.00 dollars".

In [88]:
txt = f"The price is {20 * 59} dollars"
print(txt)

The price is 1180 dollars


### Analisis

Kode tersebut menggunakan f-string untuk menyisipkan hasil perhitungan langsung ke dalam string. Ekspresi {20 * 59} akan dihitung terlebih dahulu oleh Python, kemudian hasilnya dimasukkan ke dalam teks. Output yang dihasilkan adalah "The total is 1180".

## Escape Character

In [89]:
txt = "We are the so-called \"Vikings\" from the north."

### Analisis

Kode tersebut menyimpan string yang mengandung tanda kutip ganda di dalamnya dengan menggunakan escape character (\"). Dengan cara ini, Python dapat mengenali tanda kutip tersebut sebagai bagian dari teks, bukan sebagai penutup string.

## Boolean Values

In [90]:
print(10 > 9)
print(10 == 9)
print(10 < 9)

True
False
False


### Analisis

Kode tersebut menggunakan operator perbandingan di Python untuk membandingkan nilai numerik.
Pernyataan 10 > 9 memeriksa apakah 10 lebih besar dari 9 dan menghasilkan True.
Pernyataan 10 == 9 memeriksa apakah 10 sama dengan 9 dan menghasilkan False.
Pernyataan 10 < 9 memeriksa apakah 10 lebih kecil dari 9 dan juga menghasilkan False.
Hasil dari ketiga perbandingan tersebut ditampilkan dalam bentuk nilai boolean, yaitu True atau False.

In [91]:
a = 200
b = 33

if b > a:
  print("b is greater than a")
else:
  print("b is not greater than a")

b is not greater than a


### Analisis

Kode tersebut menggunakan pernyataan if-else untuk membandingkan dua variabel, yaitu a = 200 dan b = 33. Program memeriksa apakah b lebih besar dari a. Karena 33 tidak lebih besar dari 200, kondisi if bernilai False, sehingga blok else dijalankan dan menampilkan output sesuai perintah di dalamnya.

## Evaluate Values and Variables

In [92]:
print(bool("Hello"))
print(bool(15))

True
True


### Analisis

Kode tersebut menggunakan fungsi bool() untuk mengonversi nilai menjadi tipe data boolean (True atau False) di Python. Nilai seperti string yang tidak kosong (misalnya "Hello") dan angka selain nol (misalnya 15) akan menghasilkan True, karena dianggap memiliki nilai (truthy) dalam konteks logika Python.

In [93]:
x = "Hello"
y = 15

print(bool(x))
print(bool(y))

True
True


### Analisis

Kode tersebut menggunakan fungsi bool() untuk mengonversi nilai variabel menjadi tipe data boolean (True atau False). Karena x = "Hello" merupakan string tidak kosong dan y = 15 merupakan angka selain nol, maka keduanya akan menghasilkan nilai True saat dikonversi menggunakan bool().

## Most Values are True

In [95]:
bool("abc")
bool(123)
bool(["apple", "cherry", "banana"])

True

### Analisis

Kode tersebut menggunakan fungsi bool() untuk mengonversi berbagai tipe data menjadi nilai boolean. Dalam konteks Python, string tidak kosong seperti "abc", angka selain nol seperti 123, dan daftar (list) yang berisi elemen seperti ["apple", "cherry", "banana"] semuanya dianggap True, karena memiliki nilai (truthy) ketika diuji dalam konteks boolean.

## Some Values are False

In [96]:
bool(False)
bool(None)
bool(0)
bool("")
bool(())
bool([])
bool({})

False

### Analisis

Kode tersebut menggunakan fungsi bool() untuk mengevaluasi berbagai nilai yang dianggap False dalam Python. Nilai-nilai seperti False, None, 0, string kosong "", tuple kosong (), list kosong [], dan dictionary kosong {} semuanya akan menghasilkan False, karena tidak memiliki nilai (falsy) dalam konteks boolean.

In [97]:
class myclass():
  def __len__(self):
    return 0

myobj = myclass()
print(bool(myobj))

False


### Analisis

Kode tersebut mendefinisikan kelas myclass dengan metode khusus len() yang selalu mengembalikan 0. Saat objek myobj dibuat dan dievaluasi menggunakan bool(), Python secara otomatis memanggil metode len() untuk menentukan nilai boolean objek tersebut. Karena len() mengembalikan 0, maka objek dianggap False dalam konteks boolean.

## Functions can Return a Boolean

In [98]:
def myFunction() :
  return True

print(myFunction())


True


### Analisis

Kode tersebut mendefinisikan fungsi myFunction() yang selalu mengembalikan nilai True. Saat fungsi tersebut dipanggil menggunakan print(myFunction()), Python akan mengeksekusi fungsi tersebut dan menampilkan True sebagai hasil outputnya.

In [99]:
def myFunction() :
  return True

if myFunction():
  print("YES!")
else:
  print("NO!")

YES!


### Analisis

Kode tersebut mendefinisikan fungsi myFunction() yang selalu mengembalikan True. Nilai hasil fungsi tersebut kemudian dievaluasi dalam pernyataan if. Karena hasilnya True, blok if dijalankan dan menampilkan output "YES!".

In [101]:
x = 200
print(isinstance(x, int))

True


### Analisis

Kode tersebut menggunakan fungsi isinstance() untuk memeriksa apakah variabel x = 200 bertipe int. Karena 200 merupakan bilangan bulat (integer), maka saat perintah print(isinstance(x, int)) dijalankan, hasil yang ditampilkan adalah True.

## Python Operators

In [102]:
print(10 + 5)


15


### Analisis

Kode tersebut melakukan operasi penjumlahan antara dua angka, 10 dan 5, menggunakan operator +. Perintah print(10 + 5) akan menampilkan 15.

In [103]:
sum1 = 100 + 50      # 150 (100 + 50)
sum2 = sum1 + 250    # 400 (150 + 250)
sum3 = sum2 + sum2   # 800 (400 + 400)


### Analisis

Kode tersebut melakukan beberapa operasi penjumlahan bertahap. Pertama, sum1 = 100 + 50 menghasilkan 150. Kemudian, sum2 = sum1 + 250 menghasilkan 400, dan terakhir sum3 = sum2 + sum2 menghasilkan 800. Hasilnya menunjukkan bagaimana nilai sebelumnya dapat digunakan dalam perhitungan berikutnya.

## Arithmetic Operators

In [104]:
x = 15
y = 4

print(x + y)
print(x - y)
print(x * y)
print(x / y)
print(x % y)
print(x ** y)
print(x // y)

19
11
60
3.75
3
50625
3


### Analisis

Kode ini menampilkan berbagai operasi aritmetika yang dilakukan pada variabel x = 15 dan y = 4. Hasilnya antara lain: penjumlahan (x + y) menghasilkan 19, pengurangan (x - y) menghasilkan 11, serta perkalian (x * y) memberikan hasil 60. Untuk operasi pembagian (x / y), hasilnya adalah 3.75, sementara operator modulus (x % y) memberikan sisa 3. Selanjutnya, operator perpangkatan (x ** y) menghasilkan 50625, dan pembagian bulat (x // y) menghasilkan 3. Semua hasil ini kemudian dicetak ke layar dengan fungsi print().

In [18]:
x = 12
y = 5

print(x / y)


2.4


### Analisis

Pada bagian ini, kode melakukan pembagian antara x = 12 dan y = 5 menggunakan operator /. Operator ini menghasilkan nilai dalam bentuk desimal (float). Saat dijalankan, perintah print(x / y) akan menampilkan hasil 2.4 sebagai keluaran di layar.

In [106]:
x = 12
y = 5

print(x // y)

2


### Analisis

Kode berikutnya memanfaatkan operator // untuk menghitung hasil pembagian bulat antara x = 12 dan y = 5. Operator ini hanya mengambil bagian bilangan bulat dari hasil pembagian dan mengabaikan sisa desimalnya. Karena 12 dibagi 5 sama dengan 2.4, maka hasil x // y adalah 2.

## The Walrus Operator

In [107]:
numbers = [1, 2, 3, 4, 5]
count = len(numbers)
if count > 3:
    print(f"List has {count} elements")

if (count := len(numbers)) > 3:
    print(f"List has {count} elements")

List has 5 elements
List has 5 elements


### Analisis

Kode ini menggunakan list numbers = [1, 2, 3, 4, 5] untuk menunjukkan dua cara menghitung jumlah elemen. Pertama, panjang list dihitung menggunakan len() dan disimpan ke variabel count. Kedua, digunakan walrus operator (:=) yang memungkinkan proses perhitungan dan penugasan dilakukan dalam satu langkah, seperti if (count := len(numbers)) > 3:. Kedua metode ini sama-sama menghasilkan output “List has 5 elements”, tetapi versi dengan walrus operator membuat kode lebih ringkas dan efisien.

## Comparison Operators

In [108]:
x = 5
y = 3

print(x == y)
print(x != y)
print(x > y)
print(x < y)
print(x >= y)
print(x <= y)


False
True
True
False
True
False


### Analisis

Kode ini menunjukkan cara kerja operator perbandingan antara dua variabel, x = 5 dan y = 3. Operator == memeriksa kesamaan (hasilnya False), != memeriksa ketidaksamaan (True), dan > memeriksa apakah x lebih besar dari y (True). Semua hasil ini berupa nilai boolean yang mencerminkan hasil evaluasi logika Python, dan ditampilkan menggunakan fungsi print().

## Chaining Comparison Operators

In [110]:
x = 5

print(1 < x < 10)

print(1 < x and x < 10)


True
True


### Analisis

Bagian terakhir memperlihatkan dua cara untuk memeriksa apakah nilai x = 5 berada di antara 1 dan 10. Metode pertama menggunakan chained comparison 1 < x < 10, sedangkan cara kedua memakai operator logika and dalam bentuk 1 < x and x < 10. Keduanya sama-sama menghasilkan True, karena nilai x memang berada dalam rentang tersebut. Perbandingan ini memperlihatkan bahwa Python mendukung ekspresi logika yang ringkas dan mudah dibaca.

## Logical Operators

In [111]:
x = 5

print(x > 0 and x < 10)

True


### Analisis

Kode ini memanfaatkan operator logika and untuk memeriksa apakah nilai variabel x = 5 berada dalam rentang antara 0 dan 10. Ekspresi yang digunakan adalah x > 0 and x < 10, di mana kedua kondisi harus bernilai benar (True) agar hasil keseluruhan juga True. Karena nilai 5 memang lebih besar dari 0 dan lebih kecil dari 10, maka saat kode print(x > 0 and x < 10) dijalankan, program akan menampilkan output True di layar.

In [112]:
x = 5

print(x < 5 or x > 10)

False


### Analisis

Kode ini menggunakan operator logika or untuk memeriksa apakah nilai variabel x = 5 memenuhi salah satu dari dua kondisi, yaitu x < 5 atau x > 10. Operator or akan menghasilkan nilai True jika salah satu kondisi bernilai benar. Namun, pada kasus ini, x < 5 bernilai False dan x > 10 juga False, sehingga hasil keseluruhan ekspresi menjadi False. Ketika perintah print(x < 5 or x > 10) dijalankan, program akan menampilkan False di layar.

In [113]:
x = 5

print(not(x > 3 and x < 10))

False


### Analisis

Kode ini menggunakan operator logika not untuk membalikkan hasil logika dari ekspresi x > 3 and x < 10. Pada awalnya, kondisi x > 3 and x < 10 bernilai True, karena nilai x = 5 memang lebih besar dari 3 dan lebih kecil dari 10. Namun, setelah ditambahkan operator not, hasil logika tersebut dibalik menjadi False. Oleh karena itu, saat perintah print(not(x > 3 and x < 10)) dijalankan, program akan menampilkan False di layar.

## Identity Operators

In [114]:
x = ["apple", "banana"]
y = ["apple", "banana"]
z = x

print(x is z)
print(x is y)
print(x == y)

True
False
True


### Analisis

Kode ini menjelaskan perbedaan antara operator is dan == dalam Python. Variabel z diset agar sama dengan x, sehingga keduanya mengacu pada objek yang sama di memori. Akibatnya, ekspresi x is z menghasilkan True karena identitas objeknya sama. Sementara itu, variabel y dibuat sebagai list baru dengan isi yang sama seperti x, tetapi menempati lokasi memori yang berbeda. Oleh karena itu, x is y menghasilkan False. Namun, ketika dibandingkan menggunakan ==, Python hanya memeriksa kesamaan nilai dari elemen-elemen list, bukan identitas objeknya, sehingga x == y menghasilkan True. Hal ini menunjukkan bahwa is digunakan untuk membandingkan identitas objek, sedangkan == digunakan untuk membandingkan isi atau nilai.

In [115]:
x = ["apple", "banana"]
y = ["apple", "banana"]

print(x is not y)

True


### Analisis 

Kode ini menggunakan operator identitas is not untuk memeriksa apakah dua variabel, x dan y, tidak menunjuk ke objek yang sama di memori. Meskipun kedua variabel tersebut memiliki isi yang sama, Python memperlakukan mereka sebagai objek yang berbeda karena dibuat secara terpisah. Oleh sebab itu, ekspresi x is not y menghasilkan True. Hal ini menunjukkan bahwa operator is not digunakan untuk membandingkan identitas objek, bukan kesamaan nilai di dalamnya.

## Difference Between is and ==

In [116]:
x = [1, 2, 3]
y = [1, 2, 3]

print(x == y)
print(x is y)


True
False


### Analisis

Kode ini memperlihatkan perbedaan penting antara kesamaan nilai (==) dan identitas objek (is) dalam Python. Operator == digunakan untuk membandingkan isi atau nilai dari dua objek. Karena kedua list x dan y memiliki elemen yang sama, maka ekspresi x == y menghasilkan True.
Sebaliknya, operator is digunakan untuk membandingkan identitas objek, yaitu apakah kedua variabel menunjuk ke lokasi memori yang sama. Dalam kasus ini, meskipun x dan y berisi data yang identik, keduanya merupakan objek berbeda, sehingga x is y menghasilkan False.

## Membership Operators

In [117]:
fruits = ["apple", "banana", "cherry"]

print("banana" in fruits)

True


### Analisis

Kode ini menggunakan operator keanggotaan (in) untuk memeriksa apakah elemen "banana" terdapat di dalam list fruits = ["apple", "banana", "cherry"]. Operator in akan menghasilkan True jika nilai yang dicari memang ada di dalam koleksi tersebut. Karena "banana" memang merupakan salah satu elemen dalam list fruits, maka ekspresi print("banana" in fruits) akan menampilkan True di layar.

In [118]:
fruits = ["apple", "banana", "cherry"]

print("pineapple" not in fruits)


True


### Analisis

Kode ini memanfaatkan operator keanggotaan not in untuk memeriksa apakah elemen "pineapple" tidak terdapat di dalam list fruits = ["apple", "banana", "cherry"]. Operator not in akan menghasilkan True jika nilai yang diperiksa memang tidak ada dalam koleksi. Karena "pineapple" tidak termasuk dalam daftar elemen fruits, maka ekspresi print("pineapple" not in fruits) akan menampilkan True saat dijalankan.

## Membership in Strings

In [119]:
text = "Hello World"

print("H" in text)
print("hello" in text)
print("z" not in text)

True
False
True


### Analisis

Kode ini menunjukkan bagaimana operator in dan not in digunakan untuk memeriksa keberadaan karakter atau substring di dalam sebuah string. Dengan text = "Hello World", ekspresi "H" in text menghasilkan True karena huruf H memang ada dalam string tersebut. Sebaliknya, "hello" in text menghasilkan False, sebab Python bersifat case-sensitive (membedakan huruf besar dan kecil), sehingga "hello" tidak dianggap sama dengan "Hello". Sementara itu, ekspresi "z" not in text bernilai True karena huruf z memang tidak terdapat dalam string. Hasil ini memperlihatkan bagaimana Python mengevaluasi keanggotaan substring di dalam teks berdasarkan kecocokan karakter dan sensitivitas huruf besar-kecil.

## Bitwise Operators

In [120]:
print(6 & 3)

2


### Analisis

Kode ini menggunakan operator bitwise AND (&) untuk melakukan perbandingan antar bit pada dua bilangan, yaitu 6 dan 3. Dalam bentuk biner, angka 6 direpresentasikan sebagai 110, sedangkan 3 sebagai 011. Operator & akan menghasilkan 1 hanya pada posisi bit di mana kedua bit bernilai 1. Hasilnya adalah 010 dalam biner, yang setara dengan 2 dalam desimal. Oleh karena itu, saat perintah print(6 & 3) dijalankan, program akan menampilkan 2 di layar.

In [121]:
print(6 | 3)

7


### Analisis 

Kode ini memanfaatkan operator bitwise OR (|) untuk membandingkan setiap bit dari dua bilangan, yaitu 6 dan 3. Dalam representasi biner, angka 6 ditulis sebagai 110, sedangkan 3 ditulis sebagai 011. Operator | akan menghasilkan 1 pada setiap posisi bit jika salah satu atau kedua bit bernilai 1. Hasil akhirnya adalah 111 dalam biner, yang setara dengan 7 dalam sistem desimal. Oleh karena itu, saat perintah print(6 | 3) dijalankan, program akan menampilkan 7 di layar.

In [122]:
print(6 ^ 3)

5


### Analisis

Kode ini menggunakan operator bitwise XOR (^) untuk membandingkan setiap bit dari dua bilangan, yaitu 6 dan 3. Dalam representasi biner, angka 6 ditulis sebagai 110, sedangkan 3 sebagai 011. Operator XOR akan menghasilkan 1 hanya jika salah satu bit bernilai 1 tetapi tidak keduanya. Dari proses ini, diperoleh hasil 101 dalam biner, yang setara dengan 5 dalam desimal. Oleh karena itu, saat perintah print(6 ^ 3) dijalankan, program akan menampilkan 5 di layar.

## Operator Precedence

In [123]:
print((6 + 3) - (6 + 3))

0


### Analisis

Kode ini menunjukkan bagaimana Python mengeksekusi ekspresi aritmetika bertingkat secara berurutan. Pertama, Python menghitung hasil dari 6 + 3, yang bernilai 9. Kemudian, hasil tersebut digunakan kembali dalam operasi pengurangan (9 - 9), sehingga menghasilkan 0. Saat perintah print((6 + 3) - (6 + 3)) dijalankan, program akan menampilkan 0 di layar. Contoh ini memperlihatkan bagaimana Python menyelesaikan operasi matematika dari dalam tanda kurung terlebih dahulu sebelum melanjutkan ke operasi berikutnya.

In [124]:
print(100 + 5 * 3)

115


### Analisis

Kode ini menunjukkan bagaimana Python menerapkan aturan prioritas operator (operator precedence) dalam melakukan perhitungan aritmetika. Pada ekspresi 100 + 5 * 3, Python terlebih dahulu melakukan perkalian karena operator * memiliki prioritas lebih tinggi dibandingkan +. Hasil dari 5 * 3 adalah 15, kemudian nilai tersebut ditambahkan ke 100, menghasilkan total 115. Saat perintah print(100 + 5 * 3) dijalankan, program akan menampilkan 115 di layar. Contoh ini menegaskan bahwa dalam Python, perkalian dan pembagian dieksekusi lebih dulu sebelum penjumlahan dan pengurangan, kecuali jika diubah dengan tanda kurung.

## Left-to-Right Evaluation

In [125]:
print(5 + 4 - 7 + 3)

5


### Analisis

Kode ini memperlihatkan bagaimana Python mengeksekusi operasi penjumlahan dan pengurangan secara berurutan sesuai urutan penulisan (dari kiri ke kanan). Pada ekspresi 5 + 4 - 7 + 3, Python melakukan perhitungan langkah demi langkah: pertama 5 + 4 = 9, kemudian 9 - 7 = 2, dan terakhir 2 + 3 = 5. Hasil akhir dari seluruh operasi tersebut adalah 5. Saat perintah print(5 + 4 - 7 + 3) dijalankan, program menampilkan 5 di layar. Contoh ini menunjukkan bahwa untuk operator dengan tingkat prioritas yang sama, Python mengeksekusinya secara berurutan dari kiri ke kanan.

## Python Conditions and If statements

In [1]:
a = 33
b = 200
if b > a:
  print("b is greater than a")

b is greater than a


### Analsiis

Kode ini menggunakan struktur percabangan if untuk membandingkan dua variabel, yaitu a = 33 dan b = 200. Ekspresi kondisi b > a diperiksa terlebih dahulu, dan karena nilai b lebih besar dari a, kondisi tersebut bernilai True. Akibatnya, Python mengeksekusi perintah di dalam blok if, yaitu mencetak pesan "b is greater than a" ke layar. Program ini menunjukkan penggunaan dasar dari percabangan dalam Python, di mana keputusan diambil berdasarkan hasil evaluasi logika dari suatu kondisi perbandingan.

## How If Statements Work

In [3]:
number = 15
if number > 0:
  print("The number is positive")


The number is positive


### Analisis

Kode ini mendefinisikan variabel number dengan nilai 15, kemudian menggunakan pernyataan if untuk memeriksa apakah nilai tersebut lebih besar dari 0. Kondisi number > 0 dievaluasi, dan karena bernilai True, program mengeksekusi perintah di dalam blok if, yaitu mencetak pesan "The number is positive" ke layar. Contoh ini menunjukkan penggunaan sederhana dari struktur percabangan dalam Python untuk menentukan apakah suatu bilangan bernilai positif.

## Multiple Statements in If Block

In [4]:
age = 20
if age >= 18:
  print("You are an adult")
  print("You can vote")
  print("You have full legal rights")

You are an adult
You can vote
You have full legal rights


### Analisis

Kode ini menetapkan variabel age dengan nilai 20, kemudian menggunakan pernyataan if untuk memeriksa apakah usia tersebut lebih besar atau sama dengan 18. Karena kondisi age >= 18 bernilai True, program mengeksekusi seluruh perintah di dalam blok if. Akibatnya, tiga baris pesan ditampilkan ke layar, yaitu:
"You are an adult"
"You can vote"
"You have full legal rights"
Contoh ini menunjukkan bahwa satu kondisi if dapat menjalankan lebih dari satu perintah secara berurutan, selama semua pernyataan tersebut berada di dalam blok yang sama dan syaratnya terpenuhi.

## Using Variables in Conditions

In [5]:
is_logged_in = True
if is_logged_in:
  print("Welcome back!")

Welcome back!


### Analisis

Kode ini mendefinisikan variabel is_logged_in dengan nilai True, yang menandakan bahwa pengguna telah berhasil login. Pernyataan if is_logged_in: digunakan untuk memeriksa apakah nilai variabel tersebut bernilai benar (True). Karena kondisinya terpenuhi, program mengeksekusi perintah di dalam blok if dan menampilkan pesan "Welcome back!" ke layar. Contoh ini memperlihatkan bagaimana tipe data boolean dapat digunakan untuk mengontrol alur logika program, khususnya dalam menentukan tindakan yang harus dijalankan berdasarkan kondisi tertentu.

## The Elif Keyword

In [6]:
a = 33
b = 33
if b > a:
  print("b is greater than a")
elif a == b:
  print("a and b are equal")

a and b are equal


### Analisis

Kode ini membandingkan dua variabel, a dan b, yang keduanya memiliki nilai 33. Pertama, program memeriksa kondisi if b > a: untuk mengetahui apakah b lebih besar dari a. Karena hasilnya False, Python melanjutkan ke kondisi berikutnya, yaitu elif a == b:, untuk memeriksa apakah kedua variabel memiliki nilai yang sama. Kondisi ini bernilai True, sehingga program mengeksekusi perintah di dalam blok elif dan menampilkan pesan "a and b are equal". Contoh ini menunjukkan bagaimana percabangan if–elif bekerja dengan memilih satu kondisi yang benar di antara beberapa kemungkinan yang disediakan.

## Multiple Elif Statements

In [7]:
score = 75

if score >= 90:
  print("Grade: A")
elif score >= 80:
  print("Grade: B")
elif score >= 70:
  print("Grade: C")
elif score >= 60:
  print("Grade: D")

Grade: C


### Analisis

Kode ini menggunakan struktur if–elif untuk menentukan nilai huruf berdasarkan skor yang diberikan. Variabel score bernilai 75, lalu program memeriksa setiap kondisi secara berurutan: apakah score >= 90, score >= 80, score >= 70, atau score >= 60. Karena hanya kondisi score >= 70 yang bernilai benar, program menjalankan perintah pada blok tersebut dan menampilkan "Grade: C" di layar. Program ini menunjukkan cara kerja percabangan bertingkat dalam mengelompokkan atau mengklasifikasikan data berdasarkan rentang nilai tertentu.

## How Elif Works

In [8]:
age = 25

if age < 13:
  print("You are a child")
elif age < 20:
  print("You are a teenager")
elif age < 65:
  print("You are an adult")
elif age >= 65:
  print("You are a senior")

You are an adult


### Analisis

Kode ini menggunakan struktur if–elif untuk menentukan kategori usia berdasarkan nilai variabel age, yang bernilai 25. Program memeriksa kondisi secara berurutan: apakah age < 13, age < 20, age < 65, atau age >= 65. Karena hanya kondisi age < 65 yang bernilai benar, program mengeksekusi blok tersebut dan menampilkan pesan "You are an adult" ke layar. Program ini menunjukkan cara kerja percabangan berjenjang (if–elif) dalam mengelompokkan data ke dalam kategori tertentu berdasarkan rentang nilai usia.

## When to Use Elif

In [9]:
day = 3

if day == 1:
  print("Monday")
elif day == 2:
  print("Tuesday")
elif day == 3:
  print("Wednesday")
elif day == 4:
  print("Thursday")
elif day == 5:
  print("Friday")
elif day == 6:
  print("Saturday")
elif day == 7:
  print("Sunday")

Wednesday


### Analisis

Kode ini menggunakan struktur if–elif untuk menentukan nama hari berdasarkan nilai variabel day, yang bernilai 3. Program memeriksa setiap kondisi secara berurutan mulai dari day == 1 hingga day == 7. Karena hanya kondisi day == 3 yang bernilai benar, program mengeksekusi blok tersebut dan menampilkan pesan "Wednesday" ke layar. Program ini menunjukkan cara penggunaan percabangan berjenjang (if–elif) untuk mencocokkan satu nilai dengan beberapa kemungkinan hasil.

## The Else Keyword

In [10]:
a = 200
b = 33
if b > a:
  print("b is greater than a")
elif a == b:
  print("a and b are equal")
else:
  print("a is greater than b")


a is greater than b


### Analisis


Kode ini membandingkan dua variabel, a dengan nilai 200 dan b dengan nilai 33, menggunakan struktur if–elif–else. Pertama, program memeriksa apakah b > a; kondisi ini bernilai salah. Selanjutnya, program memeriksa apakah a == b; hasilnya juga salah. Karena kedua kondisi sebelumnya tidak terpenuhi, program mengeksekusi blok else dan menampilkan pesan "a is greater than b" ke layar. Program ini menunjukkan bagaimana struktur if–elif–else digunakan untuk menentukan satu hasil akhir dari beberapa kemungkinan kondisi.

## Else Without Elif

In [12]:
a = 200
b = 33
if b > a:
  print("b is greater than a")
else:
  print("b is not greater than a")

b is not greater than a


### Analisis


Kode ini membandingkan dua variabel, a bernilai 200 dan b bernilai 33, menggunakan struktur if–else. Program terlebih dahulu memeriksa apakah b > a. Karena kondisi tersebut bernilai salah, program mengeksekusi blok else dan menampilkan pesan "b is not greater than a" ke layar. Program ini menunjukkan penggunaan percabangan sederhana untuk memilih satu dari dua kemungkinan perintah berdasarkan hasil evaluasi kondisi.

## How Else Works

In [14]:
number = 7

if number % 2 == 0:
  print("The number is even")
else:
  print("The number is odd")


The number is odd


### Analisis

Kode ini memeriksa apakah suatu bilangan merupakan bilangan genap atau bilangan ganjil dengan menggunakan operator modulus (%). Variabel number bernilai 7, kemudian program mengevaluasi kondisi number % 2 == 0. Karena hasil sisa bagi 7 dengan 2 adalah 1, kondisi tersebut bernilai False, sehingga program menjalankan blok else dan menampilkan pesan "The number is odd" ke layar. Program ini merupakan contoh sederhana penggunaan percabangan untuk menentukan jenis bilangan berdasarkan operasi aritmetika modulus.

## Complete If-Elif-Else Chain

In [15]:
temperature = 22

if temperature > 30:
  print("It's hot outside!")
elif temperature > 20:
  print("It's warm outside")
elif temperature > 10:
  print("It's cool outside")
else:
  print("It's cold outside!")

It's warm outside


### Analisis

Kode ini menggunakan struktur if–elif–else untuk menentukan kondisi cuaca berdasarkan nilai variabel temperature, yang bernilai 22. Program terlebih dahulu memeriksa apakah suhu lebih dari 30; kondisi ini bernilai False. Selanjutnya, program memeriksa apakah suhu lebih dari 20; kondisi ini bernilai True, sehingga blok elif dijalankan dan menampilkan pesan "It's warm outside" ke layar. Program ini menunjukkan bagaimana percabangan bertingkat dapat digunakan untuk mengelompokkan nilai numerik ke dalam kategori tertentu.

## Else as Fallback

In [16]:
username = "Emil"

if len(username) > 0:
  print(f"Welcome, {username}!")
else:
  print("Error: Username cannot be empty")

Welcome, Emil!


### Analisis

Kode ini digunakan untuk memeriksa apakah variabel username memiliki isi atau tidak. Variabel username bernilai "Emil", kemudian program mengevaluasi kondisi len(username) > 0, yaitu memeriksa apakah panjang teks lebih dari 0. Karena kondisi tersebut bernilai True, program menjalankan blok if dan menampilkan pesan "Welcome, Emil!" ke layar. Program ini menunjukkan cara menggunakan fungsi len() untuk memastikan input pengguna tidak kosong sebelum diproses lebih lanjut.

## Short Hand If

In [17]:
a = 5
b = 2
if a > b: print("a is greater than b")

a is greater than b


### Analisis

Kode ini membandingkan dua variabel, a = 5 dan b = 2, menggunakan bentuk singkat dari pernyataan if. Kondisi a > b bernilai True, sehingga perintah setelah tanda titik dua langsung dijalankan dan program menampilkan pesan "a is greater than b" ke layar. Program ini menunjukkan cara menggunakan single-line if statement di Python untuk menulis logika sederhana secara ringkas dalam satu baris kode.

## Short Hand If ... Else

In [18]:
a = 2
b = 330
print("A") if a > b else print("B")

B


### Analisis

Kode ini menggunakan conditional expression atau ternary operator di Python untuk menentukan hasil berdasarkan kondisi tertentu. Variabel a = 2 dan b = 330. Program memeriksa apakah a > b; karena kondisi ini False, bagian setelah else dijalankan, sehingga program menampilkan "B" ke layar. Program ini menunjukkan cara menulis percabangan sederhana dalam satu baris menggunakan format:
print("A") if condition else print("B")

## Assign a Value With If ... Else

In [19]:
a = 10
b = 20
bigger = a if a > b else b
print("Bigger is", bigger)

Bigger is 20


### Analisis

Kode ini menggunakan conditional expression atau ternary operator untuk menentukan nilai terbesar antara dua variabel. Variabel a = 10 dan b = 20. Program memeriksa kondisi a > b; karena hasilnya False, nilai b yang diambil dan disimpan dalam variabel bigger. Selanjutnya, program menampilkan pesan "Bigger is 20" ke layar. Program ini menunjukkan cara ringkas dan efisien untuk memilih salah satu dari dua nilai berdasarkan kondisi tertentu.

## Multiple Conditions on One Line

In [20]:
a = 330
b = 330
print("A") if a > b else print("=") if a == b else print("B")

=


### Analisis


Kode ini menggunakan nested conditional expression atau ternary operator bertingkat untuk membandingkan dua nilai. Variabel a = 330 dan b = 330. Program pertama memeriksa apakah a > b; karena hasilnya False, program melanjutkan ke bagian setelah else, yaitu print("=") if a == b else print("B"). Karena kondisi a == b bernilai True, program menampilkan tanda "=" ke layar. Program ini menunjukkan cara menulis logika perbandingan berantai secara ringkas dan efisien dalam satu baris kode menggunakan operator ternary.

## Practical Examples

In [21]:
x = 15
y = 20
max_value = x if x > y else y
print("Maximum value:", max_value)

Maximum value: 20


### Analisis

Kode ini menggunakan conditional expression atau ternary operator untuk menentukan nilai maksimum antara dua variabel. Variabel x = 15 dan y = 20. Program memeriksa kondisi x > y; karena hasilnya False, nilai y yang dipilih dan disimpan ke dalam variabel max_value. Selanjutnya, program menampilkan pesan "Maximum value: 20" ke layar. Program ini menunjukkan cara sederhana dan efisien untuk memilih nilai terbesar dari dua variabel hanya dengan satu baris kode.

In [22]:
username = ""
display_name = username if username else "Guest"
print("Welcome,", display_name)

Welcome, Guest


### Analisis

Kode tersebut menggunakan conditional expression untuk menentukan nama yang akan ditampilkan. Variabel username bernilai string kosong (""), yang dianggap False dalam konteks logika Python. Karena itu, ekspresi username if username else "Guest" memilih nilai "Guest" dan menyimpannya ke variabel display_name. Program kemudian mencetak “Welcome, Guest” ke layar. Program ini menunjukkan cara memberikan nilai default suatu variabel kosong menggunakan operator ternary.

## The and Operator

In [23]:
a = 200
b = 33
c = 500
if a > b and c > a:
  print("Both conditions are True")

Both conditions are True


### Analisis

Kode tersebut menggunakan operator logika and untuk memeriksa dua kondisi sekaligus. Variabel a bernilai 200, b bernilai 33, dan c bernilai 500. Program mengecek apakah a > b dan c > a. Karena kedua kondisi tersebut benar (200 > 33 dan 500 > 200), maka hasil logika and bernilai True, sehingga program jalankan perintah di dalam blok if dan mencetak “Both conditions are True” ke layar. Program ini menunjukkan cara menggabung dua syarat yang harus sama-sama benar agar suatu blok kode dijalankan.

## The or Operator

In [25]:
a = 200
b = 33
c = 500
if a > b or a > c:
  print("At least one of the conditions is True")

At least one of the conditions is True


### Analisis

Kode ini menggunakan operator logika or untuk memeriksa dua kondisi. Variabel a = 200, b = 33, dan c = 500. Program mengevaluasi apakah a > b atau a > c. Karena a > b bernilai True (200 lebih besar dari 33), seluruh ekspresi dengan or juga bernilai True, meskipun a > c salah. Akibatnya, program menjalankan perintah di dalam blok if dan mencetak "At least one of the conditions is True" ke layar. Program ini menunjukkan bahwa dengan operator or, cukup satu kondisi yang benar agar blok kode dijalankan.

## The not Operator

In [26]:
a = 33
b = 200
if not a > b:
  print("a is NOT greater than b")

a is NOT greater than b


### Analisis 

Kode ini menggunakan operator logika not untuk membalikkan hasil evaluasi kondisi. Variabel a = 33 dan b = 200. Kondisi a > b bernilai False karena 33 tidak lebih besar dari 200. Namun, karena terdapat not di depannya, hasil tersebut dibalik menjadi True, sehingga program menjalankan blok if dan menampilkan pesan "a is NOT greater than b" ke layar. Program ini menunjukkan bagaimana not digunakan untuk membalik nilai logika dari sebuah kondisi.

## Combining Multiple Operators

In [28]:
age = 25
is_student = False
has_discount_code = True

if (age < 18 or age > 65) and not is_student or has_discount_code:
  print("Discount applies!")

Discount applies!


### Analisis

Kode ini memeriksa beberapa kondisi untuk menentukan apakah diskon berlaku. Variabel age = 25, is_student = False, dan has_discount_code = True. Kondisi (age < 18 or age > 65) bernilai False, sementara not is_student bernilai True. Gabungan keduanya (False and True) menghasilkan False, tetapi karena ada or has_discount_code yang bernilai True, keseluruhan kondisi menjadi True. Akibatnya, program menampilkan "Discount applies!" ke layar. Program ini menunjukkan cara mengombinasikan operator logika and, or, dan not untuk mengevaluasi beberapa syarat sekaligus.

### Using Parentheses for Clarity

In [29]:
temperature = 25
is_raining = False
is_weekend = True

if (temperature > 20 and not is_raining) or is_weekend:
  print("Great day for outdoor activities!")

Great day for outdoor activities!


### Analisis

Kode ini memeriksa beberapa kondisi untuk menentukan apakah hari tersebut cocok untuk beraktivitas di luar ruangan. Variabel temperature = 25, is_raining = False, dan is_weekend = True. Kondisi (temperature > 20 and not is_raining) bernilai True karena suhu di atas 20 dan tidak hujan. Selain itu, is_weekend juga True, sehingga keseluruhan ekspresi dalam if bernilai True. Akibatnya, program menjalankan perintah di dalam blok dan menampilkan pesan "Great day for outdoor activities!" ke layar. Program ini menunjukkan penggunaan kombinasi operator logika untuk mengevaluasi beberapa faktor sebelum mengambil keputusan.

## More Examples

In [30]:
username = "Tobias"
password = "secret123"
is_verified = True

if username and password and is_verified:
  print("Login successful")
else:
  print("Login failed")

Login successful


### Analisis

Kode ini memeriksa proses login dengan memastikan bahwa username dan password tidak kosong serta akun sudah diverifikasi (is_verified = True). Jika semua kondisi terpenuhi, login dianggap berhasil; jika tidak, program menampilkan pesan bahwa login gagal. Meskipun sederhana, program ini memiliki kelemahan dari segi keamanan karena password disimpan dalam bentuk teks biasa (plain text) dan belum melakukan validasi terhadap database pengguna.

In [31]:
score = 85

if score >= 0 and score <= 100:
  print("Valid score")
else:
  print("Invalid score")

Valid score


### Analisis

Kode Python ini berfungsi untuk memeriksa dua kondisi menggunakan operator logika or. Variabel a bernilai 200, b bernilai 33, dan c bernilai 500. Program mengevaluasi apakah a > b atau a > c. Karena a > b bernilai True (200 lebih besar dari 33), maka keseluruhan ekspresi a > b or a > c juga bernilai True, meskipun a > c salah. Akibatnya, blok kode di dalam if dijalankan dan mencetak “At least one of the conditions is True” ke layar.

## Nested If Statements

In [32]:
x = 41

if x > 10:
  print("Above ten,")
  if x > 20:
    print("and also above 20!")
  else:
    print("but not above 20.")

Above ten,
and also above 20!


### Analisis

Kode Python ini menggunakan struktur percabangan bersarang (nested if) untuk memeriksa nilai variabel x yang bernilai 41.
Langkah pertama, program mengecek apakah x > 10. Karena kondisi tersebut benar, program mencetak “Above ten,”. Selanjutnya, di dalam blok yang sama, terdapat pengecekan kedua: apakah x > 20. Kondisi ini juga benar, sehingga program mencetak “and also above 20!”.

## How Nested If Works

In [33]:
age = 25
has_license = True

if age >= 18:
  if has_license:
    print("You can drive")
  else:
    print("You need a license")
else:
  print("You are too young to drive")

You can drive


### Analisis

Kode Python ini menggunakan percabangan bersarang (nested if) untuk menentukan apakah seseorang diperbolehkan mengemudi.
Variabel age bernilai 25 dan has_license bernilai True. Program pertama-tama memeriksa apakah age >= 18; karena kondisi ini benar, program melanjutkan ke blok if berikutnya untuk mengecek nilai has_license. Karena has_license juga True, maka program mencetak “You can drive” ke layar.

## Multiple Levels of Nesting

In [34]:
score = 85
attendance = 90
submitted = True

if score >= 60:
  if attendance >= 80:
    if submitted:
      print("Pass with good standing")
    else:
      print("Pass but missing assignment")
  else:
    print("Pass but low attendance")
else:
  print("Fail")

Pass with good standing


### Analisis

Kode Python ini menggunakan struktur if bersarang (nested if) untuk menentukan status kelulusan berdasarkan tiga kriteria: nilai (score), kehadiran (attendance), dan pengumpulan tugas (submitted).
Nilai score bernilai 85, attendance bernilai 90, dan submitted bernilai True. Program pertama memeriksa apakah score memenuhi batas kelulusan. Karena kondisinya benar, program melanjutkan untuk memeriksa attendance. Nilai kehadiran juga memenuhi syarat, sehingga program masuk ke pengecekan terakhir, yaitu submitted. Karena semua kondisi terpenuhi, program mencetak “Pass with good standing” ke layar.

## Nested If vs Logical Operators

In [35]:
temperature = 25
is_sunny = True

if temperature > 20:
  if is_sunny:
    print("Perfect beach weather!")

Perfect beach weather!


### Analisis

Kode Python ini menerapkan struktur if bersarang (nested if) untuk memeriksa dua kondisi sebelum menampilkan pesan tertentu. Variabel temperature diisi dengan nilai 25, sedangkan is_sunny bernilai True.
Pertama, program memeriksa apakah suhu (temperature) lebih besar dari 20. Karena kondisi ini benar, program masuk ke blok if berikutnya untuk mengecek apakah is_sunny juga bernilai True. Karena kedua kondisi terpenuhi, program mencetak “Perfect beach weather!” ke layar.

In [36]:
temperature = 25
is_sunny = True

if temperature > 20 and is_sunny:
  print("Perfect beach weather!")


Perfect beach weather!


### Analisis

Kode Python ini menggunakan satu pernyataan if dengan operator logika and untuk memeriksa dua kondisi sekaligus. Variabel temperature bernilai 25, sedangkan is_sunny bernilai True.
Program mengevaluasi kondisi (temperature > 20 and is_sunny). Karena suhu lebih dari 20 dan cuaca cerah, kedua kondisi bernilai True, sehingga program mengeksekusi perintah di dalam blok if dan mencetak “Perfect beach weather!” ke layar.

## More Examples

In [37]:
username = "Emil"
password = "python123"
is_active = True

if username:
  if password:
    if is_active:
      print("Login successful")
    else:
      print("Account is not active")
  else:
    print("Password required")
else:
  print("Username required")


Login successful


### Analisis

Kode Python ini menggunakan struktur if bersarang (nested if) untuk memverifikasi beberapa syarat sebelum proses login dinyatakan berhasil. Variabel username bernilai "Emil", password bernilai "python123", dan is_active bernilai True.
Program pertama-tama memeriksa apakah username memiliki isi. Jika benar, program lanjut memeriksa apakah password juga tidak kosong. Setelah kedua kondisi tersebut terpenuhi, program kemudian mengecek apakah akun pengguna aktif (is_active == True). Karena semua kondisi bernilai True, program mengeksekusi blok terdalam dan mencetak “Login successful” ke layar.

In [38]:
score = 92
extra_credit = 5

if score >= 90:
  if extra_credit > 0:
    print("A+ grade")
  else:
    print("A grade")
elif score >= 80:
  print("B grade")
else:
  print("C grade or below")

A+ grade


### Analisis 

Kode Python ini menggunakan if bersarang (nested if) untuk menentukan nilai akhir berdasarkan dua faktor, yaitu skor utama dan tambahan poin. Variabel score bernilai 92 dan extra_credit bernilai 5. Program pertama memeriksa apakah score >= 90; karena kondisi ini benar, program lanjut memeriksa apakah extra_credit > 0. Kedua kondisi terpenuhi, sehingga program mencetak “A+ grade” ke layar. Kode ini menunjukkan bagaimana percabangan bertingkat digunakan untuk menghasilkan penilaian yang lebih spesifik berdasarkan kombinasi beberapa kriteria.

## The pass Statement

In [39]:
a = 33
b = 200

if b > a:
  pass

### Analisis

Kode Python ini membandingkan dua variabel, yaitu a dengan nilai 33 dan b dengan nilai 200, menggunakan pernyataan if untuk memeriksa apakah b > a. Karena kondisi tersebut bernilai benar, program masuk ke dalam blok if. Namun, blok tersebut hanya berisi pernyataan pass, yang berfungsi sebagai placeholder agar struktur kode tetap valid tanpa mengeksekusi perintah apa pun. Akibatnya, program tidak menghasilkan output ketika dijalankan. Kode ini biasanya digunakan saat programmer ingin menyiapkan struktur logika terlebih dahulu sebelum menambahkan isi perintah di kemudian hari.

## pass in Development

In [40]:
age = 16

if age < 18:
  pass # TODO: Add underage logic later
else:
  print("Access granted")


### Analisis

Kode Python ini menggunakan struktur if–else untuk memeriksa apakah seseorang berusia di bawah 18 tahun. Variabel age bernilai 16, sehingga kondisi age < 18 bernilai benar dan program masuk ke dalam blok if. Namun, karena blok tersebut hanya berisi pernyataan pass, program tidak melakukan apa pun dan tidak menampilkan output. Komentar # TODO: Add underage logic later menunjukkan bahwa bagian ini akan diisi dengan logika tambahan di kemudian hari. Meskipun kosong, program tetap valid karena pass berfungsi sebagai placeholder untuk mencegah error akibat blok tanpa isi.

## pass vs Comments

In [41]:
score = 85

if score > 90:
  pass # This is excellent
print("Score processed")

Score processed


### Analisis

Kode tersebut mendefinisikan variabel score dengan nilai 85, lalu memeriksa apakah score > 90. Karena kondisi tersebut salah, blok if dilewati dan tidak dijalankan. Di dalam blok itu terdapat pernyataan pass, yang berfungsi sebagai placeholder agar struktur sintaks Python tetap valid meskipun belum ada aksi yang dilakukan. Setelah melewati blok if, program melanjutkan ke baris berikutnya dan mencetak “Score processed” ke layar. Kode ini menunjukkan bagaimana pass digunakan untuk mempertahankan kerangka logika program yang belum sepenuhnya diimplementasikan.

## pass with Multiple Conditions

In [42]:
value = 50

if value < 0:
  print("Negative value")
elif value == 0:
  pass # Zero case - no action needed
else:
  print("Positive value")

Positive value


### Analisis

Kode tersebut memeriksa apakah nilai value bernilai negatif, nol, atau positif. Karena value = 50, kondisi if value < 0 dan elif value == 0 bernilai salah, sehingga program masuk ke blok else dan mencetak “Positive value” ke layar. Pada bagian elif, terdapat pernyataan pass yang berfungsi sebagai placeholder, yaitu untuk menjaga struktur kode tetap valid meskipun belum ada aksi yang dilakukan. Program ini menunjukkan penggunaan pass dalam percabangan untuk menandai bagian logika yang mungkin akan diisi kemudian.

## pass in Other Contexts

In [43]:
def calculate_discount(price):
  pass # TODO: Implement discount logic

# Function exists but doesn't do anything yet

### Analisis

Kode tersebut mendefinisikan fungsi calculate_discount(price) yang belum memiliki isi. Di dalamnya terdapat pernyataan pass yang berfungsi sebagai placeholder agar fungsi tetap valid secara sintaks meskipun belum berisi logika apa pun. Komentar # TODO: Implement discount logic menunjukkan bahwa bagian tersebut akan diisi dengan perhitungan diskon pada tahap selanjutnya. Program ini berguna sebagai kerangka awal (template) saat merancang struktur fungsi sebelum implementasi logika sebenarnya ditambahkan.

## The Python Match Statement

In [45]:
day = 4
match day:
  case 1:
    print("Monday")
  case 2:
    print("Tuesday")
  case 3:
    print("Wednesday")
  case 4:
    print("Thursday")
  case 5:
    print("Friday")
  case 6:
    print("Saturday")
  case 7:
    print("Sunday")

Thursday


### Analisis

Kode tersebut menggunakan struktur match-case (fitur baru di Python 3.10 ke atas) untuk mencocokkan nilai variabel day. Variabel day bernilai 4, sehingga program menemukan kecocokan pada case 4 dan menjalankan perintah di dalamnya, mencetak “Thursday” ke layar. Program ini berfungsi mirip dengan pernyataan switch-case di bahasa pemrograman lain, dan memudahkan penulisan percabangan yang memiliki banyak kondisi.

## Default Value

In [46]:
day = 4
match day:
  case 6:
    print("Today is Saturday")
  case 7:
    print("Today is Sunday")
  case _:
    print("Looking forward to the Weekend")

Looking forward to the Weekend


### Analisis

Kode tersebut menggunakan struktur match-case untuk mencocokkan nilai variabel day. Nilai day adalah 4, sehingga tidak cocok dengan case 6 maupun case 7. Akibatnya, program mengeksekusi case _, yang berfungsi sebagai default case untuk menangani semua nilai yang tidak sesuai dengan case lain, dan mencetak “Looking forward to the Weekend” ke layar. Kode ini menunjukkan cara menggunakan match-case dengan simbol _ untuk menangani kondisi umum atau nilai yang tidak tercakup secara spesifik.

## Combine Values

In [47]:
day = 4
match day:
  case 1 | 2 | 3 | 4 | 5:
    print("Today is a weekday")
  case 6 | 7:
    print("I love weekends!")

Today is a weekday


### Analisis

Kode tersebut menggunakan struktur match-case untuk mengevaluasi nilai variabel day dan menampilkan pesan sesuai kategorinya. Variabel day bernilai 4, sehingga cocok dengan pola case 1 | 2 | 3 | 4 | 5, yang mewakili hari kerja (weekday). Akibatnya, program mengeksekusi blok tersebut dan mencetak “Today is a weekday” ke layar. Kode ini menunjukkan bagaimana beberapa nilai dapat digabungkan dalam satu case menggunakan tanda pemisah | (pipe) untuk membuat kondisi lebih ringkas dan mudah dibaca.

## If Statements as Guards

In [48]:
month = 5
day = 4
match day:
  case 1 | 2 | 3 | 4 | 5 if month == 4:
    print("A weekday in April")
  case 1 | 2 | 3 | 4 | 5 if month == 5:
    print("A weekday in May")
  case _:
    print("No match")

A weekday in May


### Analisis


Kode tersebut menggunakan struktur match-case dengan guard condition (if) untuk mencetak pesan berdasarkan nilai day dan month. Variabel month bernilai 5 dan day bernilai 4. Program pertama memeriksa case 1 | 2 | 3 | 4 | 5 if month == 4, tetapi kondisi month == 4 tidak terpenuhi sehingga dilewati. Selanjutnya, case 1 | 2 | 3 | 4 | 5 if month == 5 terpenuhi karena day cocok dan month sama dengan 5, sehingga program mencetak “A weekday in May” ke layar. Kode ini menunjukkan bagaimana guard condition digunakan dalam match-case untuk menambahkan syarat tambahan dan memperjelas logika pencocokan.